### PR TITLE
Bytes and characters

### PR DESCRIPTION
**Pull request by eggrobin**
Make Array easier to construct (as a view of a container, and from a string literal), and change a number of arrays to use `char` for UTF-8 bytes and `std::uint8_t` otherwise (we may want to switch to `std::byte` rather than `std::uint8_t` at a later date).

Note that there remain some uses of `char` for non-text bytes, to be fixed in a subsequent pull request.

### CODE REVIEW

File: base/array.hpp [hunk 1]
```diff
@@ -102,16 +130,33 @@ class BoundedArray final {
 using Bytes = Array<std::uint8_t>;
 using UniqueBytes = UniqueArray<std::uint8_t>;
 
+inline UniqueBytes SerializeAsBytes(
+    google::protobuf::MessageLite const& message);
```
Comment (pleroy) [inline base/array.hpp]: This doesn't seem to belong here.  Could we put it, e.g., in a `base/proto.hpp` or `base/serialization.hpp` header?

File: base/array_body.hpp [hunk 2]
```diff
@@ -21,14 +21,36 @@ template<typename Size, typename>
 Array<Element>::Array(Element* const data, Size const size)
     : data(data), size(static_cast<std::int64_t>(size)) {}
 
+template<typename Element>
+template<typename Container, typename>
+constexpr Array<Element>::Array(Container container)
```
Comment (pleroy) [inline base/array_body.hpp]: Why not `const&`?

File: base/base32768_body.hpp [hunk 3]
```diff
@@ -229,23 +229,22 @@ void Base32768Encode(Array<std::uint8_t const> input,
     std::int32_t code_point = (data & mask) >> shift;
     CHECK_LE(0, code_point);
     CHECK_LT(code_point, 1 << bits_per_code_point);
-    std::memcpy(
-        output.data, &(repertoire->Encode(code_point)), sizeof(char16_t));
+    *output.data = repertoire->Encode(code_point);
```
Comment (pleroy) [inline base/base32768_body.hpp]: `output.data[0]`, it's cleaner.

File: base/base32768_body.hpp [hunk 4]
```diff
@@ -229,23 +229,22 @@ void Base32768Encode(Array<std::uint8_t const> input,
...
     input_bit_index += bits_per_code_point;
     input.data += input_bit_index / bits_per_byte;
     input_bit_index %= bits_per_byte;
-    output.data += sizeof(char16_t);
+    ++output.data;
   }
 }
-UniqueArray<std::uint8_t> Base32768Encode(Array<std::uint8_t const> input,
-                                          bool const null_terminated) {
+UniqueArray<char16_t> Base32768Encode(Array<std::uint8_t const> input,
```
Comment (pleroy) [inline base/base32768_body.hpp]: `input` can be `const` here.

File: base/base32768_body.hpp [hunk 5]
```diff
@@ -297,14 +295,14 @@ void Base32768Decode(Array<std::uint8_t const> input,
     output_bit_index += bits_per_code_point;
     output.data += output_bit_index / bits_per_byte;
     output_bit_index %= bits_per_byte;
-    input.data += sizeof(char16_t);
+    ++input.data;
   }
 }
 
-UniqueArray<std::uint8_t> Base32768Decode(Array<std::uint8_t const> input) {
+UniqueArray<std::uint8_t> Base32768Decode(Array<char16_t const> input) {
```
Comment (pleroy) [inline base/base32768_body.hpp]: Same.

File: base/base32768_test.cpp [hunk 6]
```diff
@@ -14,44 +14,27 @@ namespace base {
...
     EXPECT_EQ(0,
               std::char_traits<char16_t>::compare(
-                  output32768, base32768, base32768_length));
-    for (int i = 0; i < base32768_length; ++i) {
-      EXPECT_EQ(output32768[i], base32768[i])
-          << "index=" << i << " actual=" << std::hex << output32768[i]
-          << " expected=" << base32768[i];
+                  output.data.get(), base32768.data, base32768.size));
+    for (int i = 0; i < base32768.size; ++i) {
+      EXPECT_EQ(output.data[i], output.data[i])
```
Comment (pleroy) [inline base/base32768_test.cpp]: Hmm, this will not fail often...

File: base/hexadecimal_test.cpp [hunk 7]
```diff
@@ -18,147 +18,147 @@ namespace base {
...
   UniqueBytes bytes(byte_count);
   HexadecimalDecode(digits_.get(), bytes.get());
-  EXPECT_EQ(bytes_, bytes);
+  EXPECT_EQ(bytes_, bytes.get());
 }
 
 TEST_F(HexadecimalTest, UniqueEncodeAndDecode) {
   auto const digits =
-      HexadecimalEncode(bytes_.get(), /*null_terminated=*/false);
+      HexadecimalEncode(bytes_, /*null_terminated=*/false);
```
Comment (pleroy) [inline base/hexadecimal_test.cpp]: Would fit on the previous line.

File: ksp_plugin/interface.cpp [hunk 8]
```diff
@@ -372,16 +368,13 @@ void principia__DeserializePlugin(char const* const serialization,
   }
 
   // Decode the hexadecimal representation.
-  std::uint8_t const* const hexadecimal =
-      reinterpret_cast<std::uint8_t const*>(serialization);
-  int const hexadecimal_size = serialization_size;
-  auto bytes = HexadecimalDecode({hexadecimal, hexadecimal_size});
+  auto bytes = HexadecimalDecode({serialization, serialization_size});
```
Comment (pleroy) [inline ksp_plugin/interface.cpp]: `const`.

File: ksp_plugin/interface.cpp [hunk 9]
```diff
@@ -372,16 +368,13 @@ void principia__DeserializePlugin(char const* const serialization,
...
   auto const bytes_size = bytes.size;
   (*deserializer)->Push(std::move(bytes));
 
   // If the data was empty, delete the deserializer.  This ensures that
   // |*plugin| is filled.
-  if (bytes_size == 0) {
+  if (bytes.size == 0) {
```
Comment (pleroy) [inline ksp_plugin/interface.cpp]: This is UB.  Exercise for the reader: find why.  Hint: there was a reason for the intermediate variable.

File: ksp_plugin/plugin.cpp [hunk 10]
```diff
@@ -267,15 +266,13 @@ void Plugin::EndInitialization() {
...
-          reinterpret_cast<std::uint8_t const*>(bytes.data()), bytes.size()),
-      /*null_terminated=*/true);
+  auto const hex =
+      base::HexadecimalEncode(SerializeAsBytes(ephemeris_message).get(),
+                              /*null_terminated=*/true);
   // Begin and end markers to make sure the hex did not get clipped (this might
   // happen if the message is very big).
   LOG(INFO) << "Ephemeris at initialization:\nbegin\n"
-            << reinterpret_cast<char const*>(hex.data.get()) << "\nend";
+            << hex.data.get() << "\nend";
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Would fit on the previous line.

File: base/array.hpp [hunk 11]
```diff
@@ -102,16 +130,33 @@ class BoundedArray final {
 using Bytes = Array<std::uint8_t>;
```
Comment (pleroy) [inline base/array.hpp]: Are these names still useful?  I mean, we don't have them for `char` and `char16_t`.

File: ksp_plugin/interface.cpp (hunk 8 above)
Comment (eggrobin) [reply ksp_plugin/interface.cpp]: No, moved below.

File: ksp_plugin/interface.cpp (hunk 9 above)
Comment (eggrobin) [reply ksp_plugin/interface.cpp]: Done.

File: ksp_plugin/plugin.cpp (hunk 10 above)
Comment (eggrobin) [reply ksp_plugin/plugin.cpp]: Not fully; this is how clang-format folds it.

File: base/array.hpp (hunk 11 above)
Comment (eggrobin) [reply base/array.hpp]: We could have them, if we can find good names (Array<char> is some sort of mutable string view, Array<char const> is a string view by any other name, and then there's the same in UTF-16).

File: base/hexadecimal_test.cpp (hunk 7 above)
Comment (eggrobin) [reply base/hexadecimal_test.cpp]: Done.

File: base/base32768_test.cpp (hunk 6 above)
Comment (eggrobin) [reply base/base32768_test.cpp]: Done.

File: base/base32768_body.hpp (hunk 5 above)
Comment (eggrobin) [reply base/base32768_body.hpp]: `++input.data`.

File: base/base32768_body.hpp (hunk 4 above)
Comment (eggrobin) [reply base/base32768_body.hpp]: `input.data += input_bit_index / bits_per_byte`.

File: base/base32768_body.hpp (hunk 3 above)
Comment (eggrobin) [reply base/base32768_body.hpp]: Done.

File: base/serialization.hpp [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+#include "base/array.hpp"
+
+#include "google/protobuf/message_lite.h"
+
+namespace principia {
+namespace base {
+
+inline UniqueBytes SerializeAsBytes(
+    google::protobuf::MessageLite const& message);
```
Comment (pleroy) [inline base/serialization.hpp]: Fits on the previous line?

File: base/serialization_body.hpp [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "base/serialization.hpp"
+
+namespace principia {
+namespace base {
+
+inline UniqueBytes SerializeAsBytes(
```
Comment (pleroy) [inline base/serialization_body.hpp]: You don't need `inline` in both places, I think.

File: base/base32768_body.hpp [hunk 14]
```diff
@@ -29,7 +29,7 @@ constexpr int CeilingLog2(int const n) {
 // Plane code points.
 class Repertoire {
  public:
-  virtual char16_t const& Encode(std::uint16_t  k) const = 0;
+  virtual char16_t Encode(std::uint16_t  k) const = 0;
```
Comment (pleroy) [inline base/base32768_body.hpp]: Two spaces before `k`.

File: base/hexadecimal_test.cpp [hunk 15]
```diff
@@ -18,7 +18,7 @@ namespace base {
 class HexadecimalTest : public testing::Test {
  protected:
   HexadecimalTest()
-    : bytes_("\x00\x7F\x80\xFF\x67\x68\0A\x07"),
```
Comment (pleroy) [inline base/hexadecimal_test.cpp]: How did that work?

File: base/serialization.hpp (hunk 12 above)
Comment (eggrobin) [reply base/serialization.hpp]: No.

**Pull request merged**